pkg/apis/pipeline/v1alpha1/task_validation.go

```diff
@@ -117,61 +117,65 @@ func validateResourceVariables(steps []corev1.Container, inputs *Inputs, outputs
 
 func validateVariables(steps []corev1.Container, prefix string, vars map[string]struct{}) *apis.FieldError {
 	for _, step := range steps {
-		if err := validateVariable("name", step.Name, prefix, vars); err != nil {
+		if err := validateTaskVariable("name", step.Name, prefix, vars); err != nil {
 			return err
 		}
-		if err := validateVariable("image", step.Image, prefix, vars); err != nil {
+		if err := validateTaskVariable("image", step.Image, prefix, vars); err != nil {
 			return err
 		}
-		if err := validateVariable("workingDir", step.WorkingDir, prefix, vars); err != nil {
+		if err := validateTaskVariable("workingDir", step.WorkingDir, prefix, vars); err != nil {
 			return err
 		}
 		for i, cmd := range step.Command {
-			if err := validateVariable(fmt.Sprintf("command[%d]", i), cmd, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("command[%d]", i), cmd, prefix, vars); err != nil {
 				return err
 			}
 		}
 		for i, arg := range step.Args {
-			if err := validateVariable(fmt.Sprintf("arg[%d]", i), arg, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("arg[%d]", i), arg, prefix, vars); err != nil {
 				return err
 			}
 		}
 		for _, env := range step.Env {
-			if err := validateVariable(fmt.Sprintf("env[%s]", env.Name), env.Value, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("env[%s]", env.Name), env.Value, prefix, vars); err != nil {
 				return err
 			}
 		}
 		for i, v := range step.VolumeMounts {
-			if err := validateVariable(fmt.Sprintf("volumeMount[%d].Name", i), v.Name, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].Name", i), v.Name, prefix, vars); err != nil {
 				return err
 			}
-			if err := validateVariable(fmt.Sprintf("volumeMount[%d].MountPath", i), v.MountPath, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].MountPath", i), v.MountPath, prefix, vars); err != nil {
 				return err
 			}
-			if err := validateVariable(fmt.Sprintf("volumeMount[%d].SubPath", i), v.SubPath, prefix, vars); err != nil {
+			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].SubPath", i), v.SubPath, prefix, vars); err != nil {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
-func validateVariable(name, value, prefix string, vars map[string]struct{}) *apis.FieldError {
-	if vs, present := extractVariablesFromString(value, prefix); present {
+func validateTaskVariable(name, value, prefix string, vars map[string]struct{}) *apis.FieldError {
+	return ValidateVariable(name, value, prefix, "(?:inputs|outputs).", "step", "taskspec.steps", vars)
+}
+
+func ValidateVariable(name, value, prefix, contextPrefix, locationName, path string, vars map[string]struct{}) *apis.FieldError {
```

Comment:
Doesn't look like it needs to be exported ? (except if we extract it in a `template` package)

Reply:
Yeah, that’s probably an artifact of an earlier setup. I’ll revisit.

Reply:
Ok, `ValidateVariable` does need to be exported, so it can be picked up by `validatePipelineVariable` in `pipeline_validation`, but `ExtractVariablesFromString` does not need to be exported. Fixing that.